File: SmartReceipts/PDF/TableContentRow.m
//
//  TableContentRow.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 10/07/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "TableContentRow.h"

@interface TableContentRow ()

@property (nonatomic, strong) IBOutlet UILabel *contentLabel;

@end

@implementation TableContentRow

- (void)setValue:(NSString *)value {
    if (!self.contentLabel) {
        self.contentLabel = [self subviewOfType:[UILabel class]];
    }
    [self.contentLabel setText:value];
}

- (UILabel *)subviewOfType:(Class)typeClass {
    for (UIView *sub in self.subviews) {
        if ([sub isKindOfClass:typeClass]) {
            return sub;
        }
    }

    return nil;
}

- (CGFloat)widthForValue:(NSString *)value {
    [self.contentLabel setText:value];
    CGSize fitSize = [self.contentLabel sizeThatFits:CGSizeMake(CGFLOAT_MAX, 100)];
    return fitSize.width + CGRectGetWidth(self.frame) - CGRectGetWidth(self.contentLabel.frame);
}

- (CGFloat)heightForValue:(NSString *)value usingWidth:(CGFloat)width {
    [self.contentLabel setText:value];
    CGFloat labelWidth = width - (CGRectGetWidth(self.frame) - CGRectGetWidth(self.contentLabel.frame));
    CGSize fitSize = [self.contentLabel sizeThatFits:CGSizeMake(labelWidth, CGFLOAT_MAX)];
    return fitSize.height + (CGRectGetHeight(self.frame) - CGRectGetHeight(self.contentLabel.frame));
}

@end

File: SmartReceipts/PDF/TableContentRow.swift
//
//  TableContentRow.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 01/12/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation

class TableContentRow: UIView {
    @IBOutlet private(set) var contentLabel: UILabel?
    
    func setValue(_ value: String) {
        if contentLabel == nil {
            contentLabel = subviewLabel()
        }
        contentLabel?.text = value
    }
    
    private func subviewLabel() -> UILabel? {
        for sub in subviews {
            if sub.isKind(of: UILabel.self) {
                return sub as? UILabel
            }
        }
        return nil
    }
    
    func widthFor(value: String) -> CGFloat {
        contentLabel?.lineBreakMode = .byWordWrapping
        contentLabel?.numberOfLines = 0
        contentLabel?.text = value
        let size = CGSize(width: 0, height: CGFloat.greatestFiniteMagnitude)
        let fitSize = contentLabel?.sizeThatFits(size)
        return fitSize!.width + (frame.width - contentLabel!.frame.width)
    }
    
    func heightFor(value: String, width: CGFloat) -> CGFloat {
        contentLabel?.text = value
        let labelWidth = width - (frame.width - contentLabel!.frame.width)
        let size = CGSize(width: labelWidth, height: CGFloat.greatestFiniteMagnitude)
        let fitSize = contentLabel?.sizeThatFits(size)
        return fitSize!.height + (frame.height - contentLabel!.frame.height)
    }
}
